File: src/main/java/io/nozemi/runescape/handlers/impl/ServiceHandler.java
package io.nozemi.runescape.handlers.impl;

import io.nozemi.runescape.handlers.Handler;
import io.nozemi.runescape.service.Service;
import org.springframework.stereotype.Component;

import java.util.List;
import java.util.Optional;

@Component
public class ServiceHandler implements Handler {

    private final List<Service> services;

    public ServiceHandler(List<Service> services) {
        this.services = services;
    }

    @Override
    public void initialize() {
        services.forEach(Service::setup);
        services.forEach(Service::start);
    }

    public <T extends Service> Optional<T> service(Class<? extends T> serviceType, boolean allowSubclass) {
        if (allowSubclass) {
            return (Optional<T>) services.stream().filter(s -> serviceType.isAssignableFrom(s.getClass())).findAny();
        }

        return (Optional<T>) services.stream().filter(s -> s.getClass() == serviceType).findAny();
    }
}

File: game-server/src/main/kotlin/net/dodian/runescape/handlers/ServiceHandler.kt
package net.dodian.runescape.handlers

import net.dodian.runescape.services.Service
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.stereotype.Component

@Component
class ServiceHandler @Autowired constructor(
    private val services: List<Service>
) : Handler {

    override fun initialize() {
        services.forEach { it.setup() }
        services.forEach { it.start() }
    }

    fun <T : Service> service(serviceType: Class<out T>, allowSubclass: Boolean): T? {
        return services.first { service ->
            if(allowSubclass) {
                serviceType.isAssignableFrom(service.javaClass)
            } else {
                service.javaClass == serviceType
            }
        } as T?
    }
}